README.md
# jwpqt

jwpqt is a native Linux port of JWPxp. It continues the recovered JWPce/JWPxp
release history while replacing the Win32 application layer with Qt 6.

The port is intentionally incremental. The historical sources remain at the
repository root as the behavior and format reference. New portable code lives
under `src/core`; Qt-specific application code lives under `src/qt`.

## Build

On Fedora, install the Qt 6 development package:

```sh
sudo dnf install qt6-qtbase-devel
```

Configure, build, and test out of tree:

```sh
cmake -S . -B /srv/tmp/jwpqt-build -G Ninja -DCMAKE_BUILD_TYPE=Debug
cmake --build /srv/tmp/jwpqt-build
ctest --test-dir /srv/tmp/jwpqt-build --output-on-failure
```

Run the editor, optionally opening a UTF-8 file:

```sh
/srv/tmp/jwpqt-build/src/qt/jwpqt [file]
```

## Current scope

The first native slice provides strict UTF-8 loading and saving plus a basic Qt
text window. Legacy JWP/JIS encodings, the original paragraph model,
kana-to-kanji conversion, dictionaries, lookup tools, and printing have not yet
been ported.

JWPce/JWPxp copyright and licensing notices remain in `_cpright.txt`,
`_readme.txt`, and `gnugpl.txt`.


changes.txt
1.41B->1.42:

1.  Single clicking on the large kanji window will no longer inserts the kanji,
    you have to double click.

2.  Fixed bug in the sort option for the dictionary.

3.  Fixed bug where JIS table would some times use a very large font.

4.  Fixed small formatting error in UTF-8 dictionaries.  "/" is now correctly
    turned into ", "

5.  Corrected tab order for "Are You Sure?" dialog box.

6.  Fixed bug that could crash dictionary during resize under some complicated
    procedures.  (Thanks Gernot Federspiel).

7.  Improved the WINFO utility to be very fast.  Older JWP format is no 
    longer written.  The code is still there and you can enable it by removing
    a single comment.

8.  WINFO utility was rewritten to work on an array list.  The utility is now 
    very fast.  The input format has been made more flexible.  Handling of the 
    verb iku is much improved.;

9.  When selecting text and moving up over the top of the window JWPce will no
    longer select all the bottom of the text.  (MS published procedure for the 
    mouse position is not really correct, it does not properly handle when 
    the mouse is at a negative coordinate.)  This is fixed for both lists and
    the main file.
    
10. The manual and help were revised to generally use "Windows" for Windows 
    95/98/... and "Windows CE" for CE platforms.
    
11. Help was added to the history dialog.

12. Manual and web updated to use new images from XP.

13. Fixed bug in CE color dialog was not setting the current color correctly.

14. Added ctrl+^ to toggle the input mode.  This was in the manual but never
    got added.
    
15. Corrected problem in handling cached kanji information.

16. Fixed bug in auto-cycle on the dictionary.  Pressing i would bring up
    character information.  This has been fixed.  The list control now
    requires pressing ctrl+I.    

17. Removed the setting to delete conversions.

18. Added options to select blank for kanji info you don't want to see.

19. When getting character information for kana characters JWPce will 
    display the romaji equivalents accepted.      
    
20. Fixed bug in the margins dialog box.

21. Added option to remove the kanji bar.

22. JWPce is now compiled with Visual C++.NET (VC++ 7.).

23. Most notes left in the code as undefined variables have been removed.
    These are now stored as comments using VC++ comment based task feature.
    This gets rid of warning messages during compilation, but the  comments
    are harder to find. since they can only be seen when viewing the task
    list for each file, not for the project.

24. JINDEX utility will now index EUC and UTF-8 dictionaries.  The utility 
    provides a number of options to control how the index is generated.

25. Fixed bug in searching Mixed dictionaries.  (Thanks Jean-Marc!)

26. Fixed auto-typing for dictionaries, now correctly reconizes most Mixed
    dictionaries.
    
27. Updated translation kits.

1.40B->1.41B:

1.  Fixed bug in IME support.

2.  Fixed minimization of dialog boxes.  This was messing up the display.

3.  Corrected some spelling.

4.  Fixed a problem with getting multiple search messages if you keep searching
    with the search dialog open.

5.  Cleaned up the past back functions.

6.  Added way to insert the kanji from a kanji-info window back into the file.
    This can be done by double clicking on the kanji box.

7.  Fixed bug in window placement so if you restore old placement the window will
    not grow one pixel each time.

8.  "Search" and "Find Next" are now in the popup menu associated with the list
    control.  Added keyboard shortcuts for search (Ctrl+F, Ctrl+S, and F8), to 
    match the main text window.  Added keyboard shortcuts for Find Next (Ctr+N
    and F9), again to match the main window.

9,  Corrected error message reporting for searches in lists (similar to #4).

10. Added memory to size of the info and more info dialog boxes.  This changed
    the configuration file.  I didn't want to do that but well, you know.

11. Fixed bug in editing user dictionary.

12. Fixed small bug in EDIT_list class processing.

13. Sort option has been added to the user dictionary, main dictionary, and to 
    the user kana->kanji conversions.  The one in the main dictionary is probably
    not too useful, but people have asked for it.  All of these sorts work 
    primarily on the kana.

14. Corrected the install for project files so if you choose not to install it.

15. Added install for .jfc file.  This option installs JWPce to edit (not open)
    JFC files.  Thanks for the suggestion by Gernot Federspiel.

16. Corrected for difference in XP.  JWPce trapped the mouse events to prevent 
    generation of a context menu event.  This did not work in XP so this was
    corrected.

17. Fixed bug in handling some times of clipboard imports.  Most noticeable effect
    is imports from Outlook will not have extra characters on the end.

18. UNICODETEXT is now the primary import format for clipboard and will be used
    even over the user selected format.  This format accurately represents
    all characters possible in JWPce.

19. Added ctrl+shift+A keyboard shortcut to select all in lists.  Also added
    this to the popup menu.

20. Added ctrl+I as keyboard shortcut to access character information from 
    lists.

21. Added ctrl+I (actually I) as keyboard shortcut for character information
    in the JIS table.

22. Changed the word right and left commands so that you no longer see the 
    carrot move.  This makes these function much faster (particularly on 
    Windows CE).

23. Changed the way some TrueType fonts are loaded.  This handles fonts with
    full UNICODE support better (such as Arial Unicode MS).  These fonts 
    report a large width because of some of the other languages.  JWPce will

jwpce.h
//===================================================================//
//                                                                   //
//  JWPce Copyright (C) Glenn Rosenthal, 1998-2001,2002              //
//  All rights reserved.                                             //
//                                                                   //
//===================================================================//

#define _WIN32_WINNT    0x0401                  // Necessary to get wheel-mouse defintions.

#include <windows.h>

#ifndef jwp_main_h
#define jwp_main_h

#include "jwp_wnce.h"                           // Windows CE stuff
#include "jwp_options.h"                        // Sets some major compile time options.
#include "resource.h"                           // Get the resource parameters for everyone
#ifndef WINCE
  #include <stdio.h>                            // Definition of sscanf() is here, but CE does not have this file
#endif

#ifdef WINELIB                                  // Defintions for UNIX WINELIB
  #include <stdlib.h>
  #include <ctype.h>
  #include <tchar.h>
  #include <malloc.h>
  #include <shlguid.h>
  #include <commdlg.h>
  #define stricmp strcasecmp
  #define CC_ANYCOLOR 0x100
  #define CC_SOLIDCOLOR 0x80
  #define CF_UNICODETEXT 99
#endif WINELIB

//===================================================================
//
//  Degugging stuff
//

//#define DEBUG_ROUTINES

#if (defined(_DEBUG) && !defined(DEBUG_ROUTINES)) 
#define DEBUG_ROUTINES
#endif _DEBUG && !DEBUG_ROUTINES

#ifdef DEBUG_ROUTINES
  extern void dprintf (TCHAR *format,...);
  extern void mprintf (TCHAR *format,...);
  #define DPRINTF   dprintf
  #define MPRINTF   mprintf
#else  DEBUG_ROUTINES  
  #define DPRINTF   (true ? (void) 0 : wsprintf )
  #define MPRINTF   (true ? (void) 0 : wsprintf )
#endif DEBUG_ROUTINES

//===================================================================
//
//  General definitions.
//

#define SIZE_BUFFER     512                     // Size of general working character buffer.
                                                // Version special is defined in options and includes special flags.
#define SIZE_WORKING    256                     // Size of a working string (smaller than a buffer).
#ifndef VERSION_SPECIAL
  #define VERSION_SPECIAL                       // Define the special ID if not already.  Used for special versions.
#endif  VERSION_SPECIAL
#define VERSION_STRING  TEXT("1.42") VERSION_SPECIAL  // Version ID number
#define VERSION_NAME    TEXT("JWPce ") VERSION_STRING // Version name.

                                // Input modes
#define MODE_KANJI      0       // Kanji (kana too)
#define MODE_ASCII      1       // ASCII
#define MODE_JASCII     2       // Japanese ASCII.
#define MODE_TOGGLE     10      // Toggles between ASCII and kanji.
#define MODE_CYCLE      11      // Cycles through all modes.
#define MODE_NOCHANGE   13      // Does the updage associated with a mode change but does not change the mode

                                // Windows globals.
extern HWND      main_window;   // Main window (parent).
extern HWND      file_window;   // Ediging window (main child).
extern HINSTANCE instance;      // Our instance.
extern HINSTANCE language;      // Language processor instance.
extern HMENU     hmenu;         // Our menu.
extern HMENU     popup;         // Our popup menu.

typedef short           INT16;  // Various sizes (mostly used in relation to JWP structures).
typedef unsigned short  UINT16;
typedef UINT16          KANJI;
typedef unsigned short  ushort;
typedef unsigned long   ulong;
typedef const TCHAR     tchar;

#define null            0       // Used to set handles to zero, clarifies memory refferences.

//
//  Menu defintiions, in order (need to be used in both jwpce.cpp, and jwp_file.cpp).
//
#if   defined(WINCE_PPC)
  #define MENU_FILE                         0       // Menu number for File menu.
  #define MENU_RECENT                       10      // Menu number for recent files list.
#elif defined(WINCE_HPC)
  #define MENU_FILE                         0       // Menu number for File menu.
  #define MENU_RECENT                       9       // Menu number for recent files list.
#else
  #define MENU_FILE                         0       // Menu number for File menu.
  #define MENU_RECENT                       13      // Menu number for recent files list.
#endif TARGET_TYPE

#define MENU_WINDOW                         4       // Menu number for the Window number
#define MENU_WINDOW_FILE_BASE               4       // Menu index for the first file name in the Windows menu.

//
//  Custom window message values.
//
#define WMU_SETWINDOWVALUE  (WM_USER+1)     // Used to set control value for all working windows.
#define WMU_GETWINDOWVALUE  (WM_USER+2)     // Used to get structrues behind windows controls.
#define WMU_COLORCHANGE     (WM_USER+3)     // Message from color box control indicating it is selected.
#define WMU_KANJIINFO       (WM_USER+4)     // Internal message used to requrest kanji information in mouse handler.
#define WMU_LOADCONTROL     (WM_USER+5)     // Request control to load selected data
#define WMU_JLBC_BEGIN      (WM_USER+6)     // Japanese list box control find beginning of entry.
#define WMU_JLBC_INSERT     (WM_USER+7)     // Insert data from list box to file.
#define WMU_SETTEXT         (WM_USER+8)     // Set text in a Japanese edit control
#define WMU_GETTEXT         (WM_USER+9)     // Get text from a Japanese edit/list control
#define WMU_RESETLIST       (WM_USER+10)    // Reset a Japanese List control
#define WMU_SETSELECT       (WM_USER+11)    // Set selection in Japanese List control
#define WMU_SETEXCLUDE      (WM_USER+12)    // Set exclusion file used for inserts
#define WMU_OKTODESTROY     (WM_USER+13)    // Message sent to non-modal dialog boxes before they are destroyed
#define WMU_CHARFROMLIST    (WM_USER+14)    // User is typing into the list box.
#define WMU_EDITFROMLIST    (WM_USER+15)    // User has enetered edit commands into the list box.
#define WMU_IMEFROMLIST     (WM_USER+16)    // User is typing text via the IME into the list box.
#define WMU_SETHISTORY      (WM_USER+17)    // Set the history used with a Japanese edit control
#define WMU_HISTORYLIST     (WM_USER+18)    // Generate a history list for this list.
#define WMU_SETINFOKANJI    (WM_USER+19)    // Set the character used for the kanji info (used in single window system)

//
//  Timers used.
//
#define TIMER_AUTOSCROLL    1               // Timer used for auto-scroll
#define TIMER_MOUSEHOLD     2               // Timer used for mouse hold events.

//===================================================================
//
//  Configuration defintiions
//
extern struct cfg *cfg;                     // Pointer to configuration used 
extern struct cfg  default_config;          // Default config stucture definines the default condition.

//===================================================================
//
//  Exported routines.
//
extern void add_dialog    (HWND hwnd,int closeable);    // Add a modeless dialog to the system list
extern void remove_dialog (HWND hwnd);                  // Remove a mdoeless dialog from the system list.
extern void set_mode      (int mode);                   // Sets the current edit mode.


#endif jwp_main_h

